### ISSUE TITLE
conv2d_transpose output shape more undefined than input shape

### ISSUE THREAD

**Issue by kaufManu** [tensorflow/tensorflow#8972]
I have posted [this on StackOverflow](http://stackoverflow.com/questions/43113984/output-shape-of-tf-nn-conv2d-transpose-is-entirely-undefined-even-though-only-ba) already but haven't gotten an answer yet, plus it feels like a bug very similar to https://github.com/tensorflow/tensorflow/issues/5807 which is why I'm posting it here as well.

Basically the problem is that the output shape of `tf.nn.conv2d_transpose` is entirely undefined, even if e.g. only one dimension in the input is unknown. So in the following code snippet, the shape of `out` is `[3, 10, 5, 5]` as expected if using the static shape to get the size of the first dimension. However if you use the dynamic shape (commented line in the snippet below), then the shape of `out` is `[?, ?, ?, ?]` instead of `[?, 10, 5, 5]`.

This is a problem for me because I am using `out` in a batch-normalization layer with `tf.contrib.layers.python.layers.batch_norm` for which certain dimensions must be defined.

```
import tensorflow as tf

input_ = tf.Variable(tf.random_normal([3, 10, 5, 1]))
w = tf.get_variable('w', initializer=tf.truncated_normal([3, 3, 5, 1], mean=0.0, stddev=0.01, dtype=tf.float32))

# output_shape = [tf.shape(input_)[0], 10, 5, 5]
output_shape = [input_.get_shape()[0].value, 10, 5, 5]
out = tf.nn.conv2d_transpose(input_,
                             filter=w,
                             output_shape=tf.pack(output_shape),
                             strides=[1, 1, 1, 1],
                             padding='SAME')
```

I am using TF v0.12 on Ubuntu 14.04, Python 3.5.2, installed in Anaconda environment through pip.

**Comment by asimshankar** [comment]
For your needs, a workaround would be to explicitly set the shape using `out.set_shape([None, 10, 5, 5])` so code downstream knows (from `out.shape` or `out.get_shape()`) what you wanted them to know.

I'll dig in a bit more about a more principled fix. Let me know if the workaround is sufficient for you for now.

**Comment by kaufManu** [comment]
@asimshankar Thanks for your reply - that workaround is sufficient for me at the moment. FWIW, an answer on SO pointed out that apparently also `tf.reshape(out_shape)` would be a possible workaround, as the reshape restores the static shape, see https://github.com/tensorflow/tensorflow/issues/833#issuecomment-278016198.

**Comment by asimshankar** [comment]
Glad to hear you have a workaround for now.

A more appropriate fix would be to have the shape inference function for [`Conv2DBackpropInput`](https://github.com/tensorflow/tensorflow/blob/de7a060aa76cb3ab6cec16f9402d9152e341b0af/tensorflow/core/ops/nn_ops.cc#L566) work out the shape from gradients/attrs. For this, contributions are welcome.

Alternatively, if and when we move graph construction in Python to run on top of the C API, the ShapeRefiner might be able to provide the partial shape of the input and make this shape inference function more accurate (FYI @skyewm)

**Comment by itsmeolivia** [comment]
Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.

**Comment by hannah-rae** [comment]
I am still having this issue with the current version of TensorFlow. For the code below, the print statement returns: 

(?, 7, 7, 32)
(?, ?, ?, ?)

```
h1 = tf.nn.relu(conv_transpose(z_matrix, [tf.shape(self.images)[0], 14, 14, 16], name="g_h1"))
h2 = conv_transpose(h1, [tf.shape(self.images)[0], 28, 28, 1], name="g_h2")
```
```

def conv_transpose(x, outputShape, name):
    with tf.variable_scope(name):
        print x.get_shape()
        w = tf.get_variable("w",[5,5, outputShape[-1], x.get_shape()[-1]], initializer=tf.truncated_normal_initializer(stddev=0.02))
        b = tf.get_variable("b",[outputShape[-1]], initializer=tf.constant_initializer(0.0))
        deconv_shape = tf.stack(outputShape)
        convt = tf.nn.conv2d_transpose(x, w, output_shape=deconv_shape, strides=[1,2,2,1])
        return convt`
```

**Comment by yongtang** [comment]
I think `tensor_util.constant_value_as_shape` could be used to address the issue? Created a PR #13193 for that.
